File: sams-mobile/TestApp/services/infra/alertService.js
import AsyncStorage from '@react-native-async-storage/async-storage';

class AlertService {
  constructor() {
    this.alerts = [
      {
        id: '1',
        serverId: '1',
        type: 'performance',
        title: 'High CPU Usage',
        message: 'CPU usage is above 80% for the last 5 minutes',
        severity: 'high',
        timestamp: new Date().toISOString(),
        acknowledged: false
      },
      {
        id: '2',
        serverId: '2',
        type: 'storage',
        title: 'Low Disk Space',
        message: 'Disk usage is above 90%',
        severity: 'medium',
        timestamp: new Date().toISOString(),
        acknowledged: false
      }
    ];
  }

  /**
   * Get all alerts
   * @returns {Array} Array of alerts
   */
  getAlerts() {
    return this.alerts.sort((a, b) => new Date(b.timestamp) - new Date(a.timestamp));
  }

  /**
   * Get alert by ID
   * @param {string} alertId - Alert ID
   * @returns {object|null} Alert object or null
   */
  getAlert(alertId) {
    return this.alerts.find(alert => alert.id === alertId) || null;
  }

  /**
   * Get alerts by server ID
   * @param {string} serverId - Server ID
   * @returns {Array} Array of alerts for specified server
   */
  getAlertsByServer(serverId) {
    return this.alerts.filter(alert => alert.serverId === serverId);
  }

  /**
   * Get alerts by severity
   * @param {string} severity - Alert severity
   * @returns {Array} Array of alerts with specified severity
   */
  getAlertsBySeverity(severity) {
    return this.alerts.filter(alert => alert.severity === severity);
  }

  /**
   * Get unacknowledged alerts
   * @returns {Array} Array of unacknowledged alerts
   */
  getUnacknowledgedAlerts() {
    return this.alerts.filter(alert => !alert.acknowledged);
  }

  /**
   * Create new alert
   * @param {string} serverId - Server ID
   * @param {string} type - Alert type
   * @param {string} title - Alert title
   * @param {string} message - Alert message
   * @param {string} severity - Alert severity (low, medium, high)
   * @returns {Promise<boolean>}
   */
  async createAlert(serverId, type, title, message, severity = 'medium') {
    try {
      const newAlert = {
        id: Date.now().toString(),
        serverId,
        type,
        title,
        message,
        severity,
        timestamp: new Date().toISOString(),
        acknowledged: false
      };
      
      this.alerts.push(newAlert);
      await this.saveAlerts();
      return true;
    } catch (error) {
      console.error('AlertService createAlert error:', error);
      return false;
    }
  }

  /**
   * Acknowledge alert
   * @param {string} alertId - Alert ID
   * @returns {Promise<boolean>}
   */
  async acknowledgeAlert(alertId) {
    try {
      const alert = this.getAlert(alertId);
      if (!alert) return false;

      alert.acknowledged = true;
      alert.acknowledgedAt = new Date().toISOString();
      
      await this.saveAlerts();
      return true;
    } catch (error) {
      console.error('AlertService acknowledgeAlert error:', error);
      return false;
    }
  }

  /**
   * Delete alert
   * @param {string} alertId - Alert ID
   * @returns {Promise<boolean>}
   */
  async deleteAlert(alertId) {
    try {
      this.alerts = this.alerts.filter(alert => alert.id !== alertId);
      await this.saveAlerts();
      return true;
    } catch (error) {
      console.error('AlertService deleteAlert error:', error);
      return false;
    }
  }

  /**
   * Clear all acknowledged alerts
   * @returns {Promise<boolean>}
   */
  async clearAcknowledgedAlerts() {
    try {
      this.alerts = this.alerts.filter(alert => !alert.acknowledged);
      await this.saveAlerts();
      return true;
    } catch (error) {
      console.error('AlertService clearAcknowledgedAlerts error:', error);
      return false;
    }
  }

  /**
   * Get alert statistics
   * @returns {object} Alert statistics
   */
  getAlertStats() {
    const total = this.alerts.length;
    const acknowledged = this.alerts.filter(alert => alert.acknowledged).length;
    const unacknowledged = total - acknowledged;
    
    const bySeverity = {
      low: this.alerts.filter(alert => alert.severity === 'low').length,
      medium: this.alerts.filter(alert => alert.severity === 'medium').length,
      high: this.alerts.filter(alert => alert.severity === 'high').length
    };

    return {
      total,
      acknowledged,
      unacknowledged,
      bySeverity
    };
  }

  /**
   * Save alerts to storage
   * @returns {Promise<void>}
   */
  async saveAlerts() {
    try {
      await AsyncStorage.setItem('alerts', JSON.stringify(this.alerts));
    } catch (error) {
      console.error('AlertService saveAlerts error:', error);
    }
  }

  /**
   * Load alerts from storage
   * @returns {Promise<void>}
   */
  async loadAlerts() {
    try {
      const alerts = await AsyncStorage.getItem('alerts');
      if (alerts) {
        this.alerts = JSON.parse(alerts);
      }
    } catch (error) {
      console.error('AlertService loadAlerts error:', error);
    }
  }

  /**
   * Initialize alert service
   * @returns {Promise<void>}
   */
  async initialize() {
    await this.loadAlerts();
  }
}

export default new AlertService();

File: src/backend/java/src/main/java/com/sams/service/AlertService.java
package com.sams.service;

import com.sams.entity.Alert;
import com.sams.entity.Server;
import com.sams.repository.AlertRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

/**
 * Service layer for Alert management
 */
@Service
@Transactional
public class AlertService {
    
    private static final Logger logger = LoggerFactory.getLogger(AlertService.class);
    
    @Autowired
    private AlertRepository alertRepository;
    
    /**
     * Create a new alert
     */
    public Alert createAlert(Server server, String title, String message, Alert.AlertSeverity severity) {
        logger.info("Creating alert for server {}: {}", server.getName(), title);
        
        Alert alert = new Alert();
        alert.setServer(server);
        alert.setTitle(title);
        alert.setMessage(message);
        alert.setSeverity(severity);
        alert.setStatus(Alert.AlertStatus.ACTIVE);
        alert.setTriggeredAt(LocalDateTime.now());
        
        return alertRepository.save(alert);
    }
    
    /**
     * Get all alerts with pagination
     */
    @Transactional(readOnly = true)
    public Page<Alert> getAllAlerts(Pageable pageable) {
        return alertRepository.findAll(pageable);
    }
    
    /**
     * Get active alerts
     */
    @Transactional(readOnly = true)
    public List<Alert> getActiveAlerts() {
        return alertRepository.findByStatusOrderByTriggeredAtDesc(Alert.AlertStatus.ACTIVE);
    }
    
    /**
     * Get alerts by server
     */
    @Transactional(readOnly = true)
    public List<Alert> getAlertsByServer(Server server) {
        return alertRepository.findByServerOrderByTriggeredAtDesc(server);
    }
    
    /**
     * Get alerts by severity
     */
    @Transactional(readOnly = true)
    public List<Alert> getAlertsBySeverity(Alert.AlertSeverity severity) {
        return alertRepository.findBySeverityOrderByTriggeredAtDesc(severity);
    }
    
    /**
     * Get critical alerts
     */
    @Transactional(readOnly = true)
    public List<Alert> getCriticalAlerts() {
        return getAlertsBySeverity(Alert.AlertSeverity.CRITICAL);
    }
    
    /**
     * Acknowledge an alert
     */
    public Optional<Alert> acknowledgeAlert(Long alertId) {
        logger.info("Acknowledging alert with ID: {}", alertId);
        
        return alertRepository.findById(alertId)
                .map(alert -> {
                    alert.setStatus(Alert.AlertStatus.ACKNOWLEDGED);
                    alert.setAcknowledgedAt(LocalDateTime.now());
                    return alertRepository.save(alert);
                });
    }
    
    /**
     * Resolve an alert
     */
    public Optional<Alert> resolveAlert(Long alertId) {
        logger.info("Resolving alert with ID: {}", alertId);
        
        return alertRepository.findById(alertId)
                .map(alert -> {
                    alert.setStatus(Alert.AlertStatus.RESOLVED);
                    alert.setResolvedAt(LocalDateTime.now());
                    return alertRepository.save(alert);
                });
    }
    
    /**
     * Auto-resolve old alerts (older than 24 hours and still active)
     */
    public int autoResolveOldAlerts() {
        LocalDateTime cutoff = LocalDateTime.now().minusHours(24);
        List<Alert> oldAlerts = alertRepository.findActiveAlertsOlderThan(cutoff);
        
        int resolvedCount = 0;
        for (Alert alert : oldAlerts) {
            alert.setStatus(Alert.AlertStatus.RESOLVED);
            alert.setResolvedAt(LocalDateTime.now());
            alertRepository.save(alert);
            resolvedCount++;
        }
        
        if (resolvedCount > 0) {
            logger.info("Auto-resolved {} old alerts", resolvedCount);
        }
        
        return resolvedCount;
    }
    
    /**
     * Get alert statistics
     */
    @Transactional(readOnly = true)
    public AlertStats getAlertStats() {
        long totalAlerts = alertRepository.count();
        long activeAlerts = alertRepository.countByStatus(Alert.AlertStatus.ACTIVE);
        long criticalAlerts = alertRepository.countBySeverityAndStatus(
                Alert.AlertSeverity.CRITICAL, Alert.AlertStatus.ACTIVE);
        long warningAlerts = alertRepository.countBySeverityAndStatus(
                Alert.AlertSeverity.WARNING, Alert.AlertStatus.ACTIVE);
        long infoAlerts = alertRepository.countBySeverityAndStatus(
                Alert.AlertSeverity.INFO, Alert.AlertStatus.ACTIVE);
        
        return new AlertStats(totalAlerts, activeAlerts, criticalAlerts, warningAlerts, infoAlerts);
    }
    
    /**
     * Delete resolved alerts older than specified days
     */
    public int cleanupOldResolvedAlerts(int daysOld) {
        LocalDateTime cutoff = LocalDateTime.now().minusDays(daysOld);
        List<Alert> oldResolvedAlerts = alertRepository.findResolvedAlertsOlderThan(cutoff);
        
        int deletedCount = oldResolvedAlerts.size();
        alertRepository.deleteAll(oldResolvedAlerts);
        
        if (deletedCount > 0) {
            logger.info("Cleaned up {} old resolved alerts", deletedCount);
        }
        
        return deletedCount;
    }
    
    /**
     * Create high CPU usage alert
     */
    public Alert createHighCpuAlert(Server server, double cpuUsage) {
        String message = String.format("High CPU usage detected: %.2f%%", cpuUsage);
        return createAlert(server, "High CPU Usage", message, Alert.AlertSeverity.WARNING);
    }
    
    /**
     * Create high memory usage alert
     */
    public Alert createHighMemoryAlert(Server server, double memoryUsage) {
        String message = String.format("High memory usage detected: %.2f%%", memoryUsage);
        return createAlert(server, "High Memory Usage", message, Alert.AlertSeverity.WARNING);
    }
    
    /**
     * Create disk space alert
     */
    public Alert createDiskSpaceAlert(Server server, double diskUsage) {
        String message = String.format("Low disk space: %.2f%% used", diskUsage);
        Alert.AlertSeverity severity = diskUsage > 90 ? Alert.AlertSeverity.CRITICAL : Alert.AlertSeverity.WARNING;
        return createAlert(server, "Disk Space Warning", message, severity);
    }
    
    /**
     * Inner class for alert statistics
     */
    public static class AlertStats {
        private final long totalAlerts;
        private final long activeAlerts;
        private final long criticalAlerts;
        private final long warningAlerts;
        private final long infoAlerts;
        
        public AlertStats(long totalAlerts, long activeAlerts, long criticalAlerts, 
                         long warningAlerts, long infoAlerts) {
            this.totalAlerts = totalAlerts;
            this.activeAlerts = activeAlerts;
            this.criticalAlerts = criticalAlerts;
            this.warningAlerts = warningAlerts;
            this.infoAlerts = infoAlerts;
        }
        
        // Getters
        public long getTotalAlerts() { return totalAlerts; }
        public long getActiveAlerts() { return activeAlerts; }
        public long getCriticalAlerts() { return criticalAlerts; }
        public long getWarningAlerts() { return warningAlerts; }
        public long getInfoAlerts() { return infoAlerts; }
    }
}
